servant/src/Servant/API/TypeLevel.hs

```diff
@@ -0,0 +1,123 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE KindSignatures #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+{-# LANGUAGE UndecidableInstances #-}
+module Servant.API.TypeLevel where
+
+import GHC.Exts(Constraint)
+import Servant.API.Capture ( Capture, CaptureAll )
+import Servant.API.ReqBody ( ReqBody )
+import Servant.API.QueryParam ( QueryParam, QueryParams, QueryFlag )
+import Servant.API.Header ( Header )
+import Servant.API.Verbs ( Verb )
+import Servant.API.Sub ( type (:>) )
+import Servant.API.Alternative ( type (:<|>) )
+
+-- * API predicates
+
+-- | Flatten API into a list of endpoints.
+type family Endpoints api where
+  Endpoints (a :<|> b) = AppendList (Endpoints a) (Endpoints b)
+  Endpoints (e :> a)   = MapSub e (Endpoints a)
+  Endpoints a = '[a]
+
+-- ** Lax inclusion
+
+-- | You may use this type family to tell the type checker that your custom
+-- type may be skipped as part of a link. This is useful for things like
+-- @'QueryParam'@ that are optional in a URI and do not affect them if they are
+-- omitted.
+--
+-- >>> data CustomThing
+-- >>> type instance IsElem' e (CustomThing :> s) = IsElem e s
+--
+-- Note that @'IsElem'@ is called, which will mutually recurse back to @'IsElem''@
```

Comment:
@jkarni: I think you misunderstood my comment. The haddock comment seems to be saying that `IsElem'` is implemented in terms of `IsElem`. (Which then in turn calls back into `IsElem'`.) But I don't see where this call is being made.